Restore: Move to trash button in Document settings

jorgefilipecosta opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/65023

This PR readds a big "Move to trash" button that was available on older releases and was meanwhile removed.

In progress:

Needs to call the action that already has a modal instead of implementing its own confirmation modal.
Should use eligibility checks of the action.
Should not appear when editing the template of a post on the post editor.


## Screenshots or screencast <!-- if applicable -->
<img width="350" alt="Screenshot 2024-09-05 at 11 51 55" src="https://github.com/user-attachments/assets/6dff25d4-475f-4fbb-9a94-c9679e52ac8a">

jorgefilipecosta committed:
avoid calling usePostActions on trash eligibility check (+2 squashed commits)
Squashed commits:
[1163143a24] Refact to be action based and change gap
[81d9da34b1] Re Add: Move to trash button.

jasmussen commented:
Popular button, let's restore it. 

We should change the spacing a little bit, make the space above trash button a bit larger. Technically it's correct at 4px:
![Screenshot 2024-09-05 at 13 51 40](https://github.com/user-attachments/assets/bc12689c-ceaa-4dd2-b3cc-d9bf82e197c1)

But optically it will feel more balanced if it uses this gap:
![Screenshot 2024-09-05 at 13 51 54](https://github.com/user-attachments/assets/5ca365f5-9f97-44c0-94ea-173e97c4ba90)

And so either you can move that button out of the _list_ so it gets the correct 16px gap between sections, or you can just add a 12px top margin to the button:

![Screenshot 2024-09-05 at 13 53 50](https://github.com/user-attachments/assets/2eb44cd6-52f2-4128-962c-60207a109d38)

jasmussen reviewed (approved):
Would love a gut-check by @jameskoster, otherwise, 👍 👍

youknowriad commented on packages/editor/src/components/sidebar/post-summary.js:
```diff
@@ -86,6 +87,7 @@ export default function PostSummary( { onActionPerformed } ) {
 										<PostsPerPage />
 										<SiteDiscussion />
 										<PostFormatPanel />
+										<PostTrash />
```
So I'm guessing @jasmussen is suggesting we move it outside of the VStack before the fills

jasmussen replied:
That would do it.

jameskoster replied:
Yeah let's move it outside of the VStack so it inherits the `16px` gap.

Mamaduka replied:
The `PostTrashCheck` wrapper is missing here.

youknowriad commented on packages/editor/src/components/post-trash/check.js:
```diff
@@ -18,29 +21,38 @@ import { store as editorStore } from '../../store';
+			permissions,
+		};
+	}, [ item, permissions ] );
+	const actions = usePostActions( { postType } );
```
It feels way too heavy for me to call usePostActions just to check if one single action is illigible or not. I'd rather just duplicate the check.

jorgefilipecosta replied:
Hi @youknowriad, I updated the code and now we don't call usePostActions on eligibility check.

youknowriad commented on packages/editor/src/components/post-trash/index.js:
```diff
@@ -1,71 +1,84 @@
 /**
  * WordPress dependencies
  */
-import { __ } from '@wordpress/i18n';
```
Why did we refactor this component? What was wrong with the previous implementation?

jorgefilipecosta replied:
We should not have two different modals to trash, we should rely on the one the actions already offer.
E.g: less strings to translate. The UI when pressing trash on actions and button also becomes the same. It is unexpected to get one UI if I press it in a given way and different UI if I press it in other way.

jorgefilipecosta replied:
This update also handles cases where a plugin removes an action, such as the move to trash action. With these changes, we also don't render this button if the action is removed.

Mamaduka replied:
I also don't see a reason for this change. Original components are well-tested, and we're not introducing any new code, so "duplication" is justified.

If the modals don't match, we can just update the confirmation modal in the `PostTrash` component.

jorgefilipecosta replied:
@Mamaduka as referred above we have API's to remove actions `unregisterEntityAction`, so I expect that if I unregister the move to trash action this button goes away. It becomes inconsistent that the actions go away on the actions menu but not on this button. If we need to retrieve the action for this check why not just use the modal the action provides?

youknowriad replied:
I agree that reuse the actions would be good, but my opinion is that the way we're reusing them now (with an adhoc component within the editor package and not exported from the dataviews package) is not good.

I think Actions are a dataviews API and should ideally not be used in other things that are not dataviews components.

Mamaduka replied:
@jorgefilipecosta, the `unregisterEntityAction` controls one part of the UI - entity actions. I don't expect them to affect other parts.

E.g., If someone removes the "Rename" action, should we also hide the Post Title field in the canvas?

jorgefilipecosta replied:
Fair points, I followed the suggestions and updated to keep the previous logic, but changed the modal to the be similar to the actions one.

jorgefilipecosta commented:
Hi @jasmussen I moved the button outside of the VStack before the fills as suggested.

@Mamaduka @youknowriad I refactored the button and checks to be based on the actions, it calls the same as the action and the check is based on action.isEligible.

Mamaduka commented:
> This update also handles cases where a plugin removes an action, such as the move to trash action. With these changes, we also don't render this button if the action is removed.

These components are public API. I think the most straightforward approach here is to restore functionality using their original code.

jorgefilipecosta committed:
don't use the action mechanism make the modals match

jorgefilipecosta committed:
update checks

jorgefilipecosta commented:
Hi @youknowriad, @Mamaduka I updated the code to following your feedback.

Mamaduka commented on packages/editor/src/components/post-trash/check.js:
```diff
@@ -34,10 +34,14 @@ export default function PostTrashCheck( { children } ) {
+			canTrashPost:
+				( ! isNew || postId ) &&
+				canUserDelete &&
+				postType !== 'wp_template' &&
+				postType !== 'wp_block' &&
+				postType !== 'wp_template_part',
```
I think `wp_navigation` should also be here. You could reuse `GLOBAL_POST_TYPES` from `store`.

https://github.com/WordPress/gutenberg/blob/57c59d1109f8c475eb2899dd7a4b1d5061a1df69/packages/editor/src/store/constants.ts#L32-L35

Mamaduka reviewed (approved):
Thank you, @jorgefilipecosta!

Left a small note regarding the `wp_navigation` post type, but besides that, everything works as expected!

jorgefilipecosta committed:
update check
